File: leetcode/graph/leetcode1368.java
package graph;

import java.util.*;

public class leetcode1368 {

    public static void main(String[] args) {
        int[][] grid = {
            {1, 1, 3},
            {3, 2, 2},
            {1, 1, 4}
        };
        leetcode1368 solution = new leetcode1368();
        int result = solution.minCost(grid);
        System.out.println(result);
    }

    public int minCost(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        Deque<int[]> queue = new ArrayDeque<>();
        Map<String, Integer> map = new HashMap<>();
        queue.offer(new int[]{0, 0, 0});
        map.put("0,0", 0);

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int r = curr[0];
            int c = curr[1];
            int cost = curr[2];

            for (int i = 0; i < 4; i++) {
                int row = r + directions[i][0];
                int col = c + directions[i][1];
                int newCost = cost;
                if (i + 1 != grid[r][c]) {
                    newCost += 1;
                }
                if (row < 0 || row >= n || col < 0 || col >= m) {
                    continue;
                }
                String key = row + "," + col;
                if (newCost >= map.getOrDefault(key, Integer.MAX_VALUE)) {
                    continue;
                }
                map.put(key, newCost);
                if (i + 1 == grid[r][c]) {
                    queue.offer(new int[]{row, col, newCost});
                } else {
                    queue.addFirst(new int[]{row, col, newCost});
                }
            }
        }
        return map.getOrDefault((n - 1) + "," + (m - 1), 0);
    }
}

File: leetcode/graph/leetcode1368.py
from collections import deque

class Solution:
    
    def minCost(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        directions = {1:[0, 1], 2:[0, -1], 3:[1, 0],4:[-1, 0]}
        
        queue = deque(([(0,0,0)]))
        n,m = len(grid),len(grid[0])
        
        min_cost =  {(0,0):0}
        
        while queue:
            r,c,cost = queue.popleft()
            if (r,c) == (n-1,m-1):
                return cost
            
            for d in directions:
                dr,dc = directions[d]
                nr,nc = r+dr, c+dc
                new_cost = cost if d==grid[r][c] else cost+1
                
                if (
                    nr<0 or nc<0 or nr==n or nc==m or new_cost>=min_cost.get((nr,nc),float("inf"))
                ):
                    continue
                min_cost[(nr,nc)] = new_cost
                if d==grid[r][c]:
                    queue.appendleft((nr,nc,new_cost))
                else:
                    queue.append((nr,nc,new_cost))